node/src/main.rs

```diff
@@ -39,14 +45,47 @@ use graph_server_websocket::SubscriptionServer as GraphQLSubscriptionServer;
 use graph_store_postgres::{Store as DieselStore, StoreConfig};
 
 fn main() {
+    let (shutdown_sender, shutdown_receiver) = oneshot::channel();
+    // Register guarded panic logger which ensures logs flush on shutdown
     let (panic_logger, _panic_guard) = guarded_logger();
-    register_panic_hook(panic_logger);
-    tokio::run(future::lazy(async_main))
+    register_panic_hook(panic_logger, Mutex::new(Some(shutdown_sender)));
+
+    // Create components for tokio context: multi-threaded runtime, reactor reference,
+    // executor context on the runtime, and Timer handle.
+    let runtime = tokio::runtime::Runtime::new().expect("Failed to create runtime");
+    let reactor = tokio_reactor::Handle::current();
+    let mut executor = runtime.executor();
+    let mut enter = tokio_executor::enter()
+        .expect("Failed to enter runtime executor, multiple executors at once");
+    let timer = Timer::default();
+    let timer_handle = timer.handle();
+
+    // Shutdown the runtime after a panic
+    std::thread::spawn(|| {
+        shutdown_receiver
+            .wait()
+            .map(|_| {
+                let _ = runtime
+                    .shutdown_now()
+                    .wait()
+                    .expect("Failed to shutdown Tokio Runtime");
+            }).expect("Runtime shutdown process did not finish");
+    });
+
+    // Setup runtime context with defaults and run the main application
+    tokio_reactor::with_default(&reactor, &mut enter, |enter| {
+        tokio_executor::with_default(&mut executor, enter, |enter| {
+            tokio_timer::with_default(&timer_handle, enter, |enter| {
```

Comment:
Is it necessary to start all this defaults? I think in tests we don't do this and just call `runtime.block_on` directly and things still work.

Reply:
I've removed the reactor default.  Removing either of the other defaults results in errors such as below.

```
`CRIT called `Result::unwrap()` on an `Err` value: SpawnError { is_shutdown: true }, location: libcore/result.rs:1009` 
```